### PR TITLE
Backports for Julia 1.4.1

### LINKED ISSUE #35201: Parsing problem with nested splatting named tuple constructor in a function call
In the code below definitions `f1` and `f2` should be equivalent I think.
However, `f1` is a syntax error, while `f2` works as expected.

```
julia> h(x; k=1) = (x, k)
h (generic function with 1 method)

julia> f1(c) = h((;c...), k=true)
ERROR: syntax: "..." expression outside call
Stacktrace:
 [1] top-level scope at REPL[8]:1

julia> f2(c) = (nt=(;c...); h(nt, k=true))
f2 (generic function with 1 method)

julia> f2(Dict(:a=>1, :b=>2))
((a = 1, b = 2), true)
```

It seems that a parser does not correctly recognize that `...` in this case is used in a nested `NamedTuple` constructor call within a call to `h`.

The problem seems to be present only if the call involves keyword arguments, so e.g.
```
julia> f1(c) = h((;c...), true)
f1 (generic function with 1 method)
```
parses correctly.

Also
```
julia> f1(c) = h(;c..., k=true)
f1 (generic function with 1 method)
```
parses, but it gives a different definition of course.

### LINKED ISSUE #35225: issubset is wrong for empty ranges
In Julia 1.4.0,
```
julia> issubset(2:1,3:4)
false
```
although of course
```
julia> issubset(collect(2:1),collect(3:4))
true
```

### LINKED ISSUE #35272: reverse searchsorted is inconsistent on 1.3
```julia
julia> searchsortedfirst(10:-1:1, 9.5; rev=true)
1

julia> searchsortedfirst(collect(10:-1:1), 9.5; rev=true)
2

julia> searchsortedlast(10:-1:1, 9.5; rev=true)
2

julia> searchsortedlast(collect(10:-1:1), 9.5; rev=true)
1
```

The range behaviour seems like the correct version, but needless to say it's extremely confusing getting different answers for these.

> searchsortedfirst(a, x; by=<transform>, lt=<comparison>, rev=false)

 > Return the index of the first value in a greater than or equal to x,
  according to the specified order. Return length(a) + 1 if x is greater than
  all values in a. a is assumed to be sorted.

It gets worse if you use all floats, now both the range and the array are wrong:

```julia
julia> searchsortedfirst(10.0:-1.0:1.0, 9.5; rev=true)
2

julia> searchsortedfirst(collect(10.0:-1.0:1.0), 9.5; rev=true)
2

julia> searchsortedlast(10.0:-1.0:1.0, 9.5; rev=true)
1

julia> searchsortedlast(collect(10.0:-1.0:1.0), 9.5; rev=true)
1
```

```julia
julia> versioninfo()
Julia Version 1.3.0
Commit 46ce4d7933 (2019-11-26 06:09 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-5300U CPU @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, broadwell)
Environment:
  JULIA_NUM_THREADS = 3
```

### LINKED ISSUE #35295: clipboard() functions error on Julia 1.4 Linux
Edit: Found this issue for clipboard("x") https://github.com/JuliaLang/julia/issues/14510, removed duplicate parts

`clipboard()` errors on Julia 1.4 (fine on 1.3)

```
julia> clipboard()
ERROR: UndefVarError: _clipboardcmds_paste not defined
Stacktrace:
 [1] clipboard() at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/InteractiveUtils/src/clipboard.jl:70
 [2] top-level scope at REPL[2]:1
```

```
Julia Version 1.4.0
Commit b8e9a9ecc6 (2020-03-21 16:36 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6700T CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
Environment:
  JULIA_EDITOR = vim
```

### LINKED ISSUE #35305: String representation of expressions with quotes in non-standard string literals
The string representation of an expression that contains quotes `"` in triple-quoted non-standard string literals (or escaped quotes in quoted non-standard string literals) is not faithfully any more. The resulting text does not represent valid Julia code and cannot be `include`d again. in Julia v1.4

```
julia> print(:(r = r"""45 " """))
r = r"45 " "
```

```
julia> print(:(r = r"45 \" "))
r = r"45 " "
```

This does not happen for "normal" strings
```
julia> print(:(r = "45 \" "))
r = "45 \" "
```

and also used to work in Julia v1.3.1. See also https://discourse.julialang.org/t/string-representation-of-expression/36729/3?u=pfarndt

### LINKED ISSUE #35391: GC.@preserve does not work within a macro in Julia 1.4
As noted on Julia Slack #internals - https://julialang.slack.com/archives/C688QKS7Q/p1586239098050400.

Macro which used to work pre Julia-1.4 stopped working. Setting aside the useless invocation of `GC.@preserve`:

```julia
julia> macro transmute(tp, ptr, off)
            :(begin
                     v = $(esc(ptr))
                     GC.@preserve v unsafe_load(convert(Ptr{$(esc(tp))}, v+$(esc(off))), 1)
             end)
         end
julia> @transmute(Int32, pointer(xs), 0)
ERROR: syntax: invalid syntax (escape (outerref Int32))
```

`@macroexpand` result is different to 1.3 - doesn’t expand the `GC.@preserve` call which “loses” `v`
1.4:
```julia
julia> @macroexpand @transmute(Int32, pointer(xs), 0)
quote
    #= REPL[344]:3 =#
    var"#2452#v" = pointer(xs)
    #= REPL[344]:4 =#
    $(Expr(:gc_preserve, :(unsafe_load(convert(Ptr{$(Expr(:escape, :Int32))}, v + $(Expr(:escape, 0))), 1)), :v))
end
```
1.3:
```julia
julia> @macroexpand @transmute(Int32, pointer(xs), 0)
quote
    #= REPL[1]:3 =#
    var"#1#v" = pointer(xs)
    #= REPL[1]:4 =#
    begin
        #= gcutils.jl:90 =#
        var"#2###422" = $(Expr(:gc_preserve_begin, Symbol("#1#v")))
        #= gcutils.jl:91 =#
        var"#3###423" = Main.unsafe_load(Main.convert(Main.Ptr{Int32}, var"#1#v" + 0), 1)
        #= gcutils.jl:92 =#
        $(Expr(:gc_preserve_end, Symbol("##422")))
        #= gcutils.jl:93 =#
        var"#3###423"
    end
end
```
Haven't dug deeper but seems like this shouldn't have broken.

### PR DESCRIPTION
**Pull request by KristofferC**
Backported PRs:
- [x] JuliaLang/julia#34760 - Add AddrSpacePtr type
- [x] JuliaLang/julia#34906 - Fix regression in inlining of invoke
- [x] JuliaLang/julia#34872 - Fix inverse of SVD of complex matrix (#34866)
- [x] JuliaLang/julia#34986 - Sockets: fix return value of getpeername/getsockname
- [x] JuliaLang/julia#34989 - Rebase of JuliaLang/julia#28526 on master: Sparse cholesky solve with sparse RHS
- [x] JuliaLang/julia#35006 - walkdir: avoid symlink loops when `follow_symlinks == false`
- [x] JuliaLang/julia#35064 - Support git credential helper resetting
- [x] JuliaLang/julia#35136 - fix bug in t-function for `>:`
- [x] JuliaLang/julia#35144 - speed up llvmcall unique name generation
- [x] JuliaLang/julia#35184 - add compat annotation for map!(f, values(dict)) (from JuliaLang/julia#31223)
- [x] JuliaLang/julia#35192 - Remove outdated paragraph from deepcopy docstring
- [x] JuliaLang/julia#35202 - fix JuliaLang/julia#35201, syntax error with named splatting inside function call
- [x] JuliaLang/julia#35227 - Bump Documenter to 0.24.7
- [x] JuliaLang/julia#35244 - fix JuliaLang/julia#35225, issubset on empty ranges
- [x] JuliaLang/julia#35180 - Linux: Remove `--nodetach` option for `xsel` copy.
- [x] JuliaLang/julia#35310 - fix JuliaLang/julia#35295, typo in change to `clipboard` on linux
- [x] JuliaLang/julia#35356 - Update `UnicodeData` checksum
- [x] JuliaLang/julia#35298 - Fix offset when widening collections
- [x] JuliaLang/julia#35315 - fix method errors printing for missing
- [x] JuliaLang/julia#35309 - fix JuliaLang/julia#35305, need escaping when printing string macro calls
- [x] JuliaLang/julia#35239 - fix a bug preventing inlining of getindex of const globals
- [x] JuliaLang/julia#35300 - [LLVM8] add patches for D75072 and D65174
- [x] JuliaLang/julia#35394 - fix JuliaLang/julia#35391, macro hygiene bug in gc_preserve
- [x] JuliaLang/julia#35387 - [LateGCLowering] Fix skipped Select lifting
- [x] JuliaLang/julia#35435 - [backport JuliaLang/julia#35274] fix JuliaLang/julia#35272, searchsorted(3:-1:1, 2.5, rev=true)

Need manual backport:
- [ ] JuliaLang/julia#34935 - [WIP] Bump SuiteSparse to 5.4.0+6

Non-merged PRs with backport label:
- [ ] JuliaLang/julia#35232 - LibGit2: add resolve_url to RemoteCallbacksStruct for LibGit2 0.99.0

### CODE REVIEW

**Comment by maleadt** [comment]
Could https://github.com/JuliaLang/julia/pull/34760 be considered? It is technically a new feature, but it's unexported, completely unused by Base, but greatly simplifies quite some code in CUDAnative. Should be a clean backport too.

File: base/gcutils.jl [hunk 1]
```diff
@@ -83,11 +83,50 @@ enable(on::Bool) = ccall(:jl_gc_enable, Int32, (Int32,), on) != 0
...
+
+When passing pointers to `ccall`, the pointed-to object is implicitly used and
+should be preserved. (Note however that you should normally just pass `x`
+directly to `ccall` which counts as an explicit use.)
+
+```jldoctest
+julia> let
+           x = "Hello"
+           p = pointer(x)
+           GC.@preserve x @ccall strlen(p::Cstring)::Cint
```
Comment (fredrikekre) [inline base/gcutils.jl]: `@ccall` is not  in 1.4.

Comment (KristofferC) [reply base/gcutils.jl]: Oops, let's drop that commit then. Good catch.

**Comment by KristofferC** [comment]
@staticfloat, do you need to clear out some cache on the buildbot for the checksum of the unicode file to update (see CI errors).

**Comment by staticfloat** [comment]
Weird, I cleared this out.  I wonder how they got the old file again?  That doesn't seem right.

You can clear them yourself by running the `nuke` jobs on buildbot, e.g. go here and click the `nuke` button in the top right: https://build.julialang.org/#/builders/139. It will queue a cleanup job that will run on every buildbot, and then you can trigger a build after that cleanup by just re-running the failing CI jobs.

**Comment by KristofferC** [comment]
I see that you ran the nuke build and I restarted the workers but doesn't seem to help.

**Comment by staticfloat** [comment]
Ah, I see the problem.  We have two branches that are downloading two different versions of UnicodeData with the same name.  How did this ever work?

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":release-1.4")`

**Comment by KristofferC** [comment]
@nanosoldier `runtests(ALL, vs = ":release-1.4")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/35229#issuecomment-610287648) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/029bcf2b4aa50c5f262521292612e901f05ee270/571c4c4_vs_b8e9a9e/report.md). cc @ararslan

**Comment by KristofferC** [comment]
@nanosoldier `runtests(ALL, vs = ":release-1.4")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/35229#issuecomment-610822494):
```
NanosoldierError: failed to run tests: AssertionError: Cannot compare identical Julia builds
```
Logs and partial data can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/5a20f219087cd591cc3204deee719e64e51a5e5d/pkgeval/by_hash/c53191a_vs_b8e9a9e/report.md)
cc @maleadt

**Comment by maleadt** [comment]
Not sure what went wrong here, so I added some debug logging.
@nanosoldier `runtests(ALL, vs = ":release-1.4")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/35229#issuecomment-610894347):
```
NanosoldierError: failed to run tests: UndefVarError: cfg not defined
```
Logs and partial data can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/0b83b2773439822acb1a75d14da75eaf88d6c12e/pkgeval/by_hash/c53191a_vs_b8e9a9e/report.md)
cc @maleadt

**Comment by maleadt** [comment]
@nanosoldier `runtests(ALL, vs = ":release-1.4")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/35229#issuecomment-610896217):
```
NanosoldierError: failed to run tests: AssertionError: Cannot compare identical Julia builds
```
Logs and partial data can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/f9e89d190be563ef6917a7614cc6d4973304cb5b/pkgeval/by_hash/c53191a_vs_b8e9a9e/report.md)
cc @maleadt

**Comment by maleadt** [comment]
I think I know what's happening:
@nanosoldier `runtests(ALL, vs = ":release-1.4")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/35229#issuecomment-610901395) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/824edc01bd2e9b62e669b7062006196fa793c342/pkgeval/by_hash/c53191a_vs_b8e9a9e/report.md). cc @maleadt

**Comment by KristofferC** [comment]
@nanosoldier `runtests(["AugmentedGaussianProcesses", "CombineML", "ControlSystemIdentification", "DiffEqGPU", "EfficientGlobalOptimization", "FunctionOperators", "ImageDistances", "InfiniteLinearAlgebra", "InterProcessCommunication", "SpatialJackknife", "Unicode_Entities"], vs = ":release-1.4")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/35229#issuecomment-611357306) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/591e1aaf0dad689371d2d0c7400e94350525774e/pkgeval/by_hash/c53191a_vs_b8e9a9e/report.md). cc @maleadt

**Comment by KristofferC** [comment]
We bumped LLVM and OpenBLAS so better rerun all the checks...

@nanosoldier `runbenchmarks(ALL, vs = ":release-1.4")`

**Comment by KristofferC** [comment]
@nanosoldier `runtests(ALL, vs = ":release-1.4")`

**Comment by rfourquet** [comment]
When is planned the release of 1.4.1? I will try before then to make a 1.4 version of #35274 which fixes a bug.

**Comment by KristofferC** [comment]
> When is planned the release of 1.4.1? 

As soon as this branch looks ready.,

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/35229#issuecomment-611968111) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/a395a55cf78e43cad56adda77020f75d03401267/3fd5790_vs_b8e9a9e/report.md). cc @ararslan

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/35229#issuecomment-611968237) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/2e37f82c612a869c8f15976e33dbe0a011117e05/pkgeval/by_hash/44fa722_vs_b8e9a9e/report.md). cc @maleadt

**Comment by KristofferC** [comment]
@nanosoldier `runtests(["ApproxBayes", "Autologistic", "BlobTracking", "CombineML", "DiffEqGPU", "HELICS", "MAT", "MusicXML", "OhMyREPL"], vs = ":release-1.4")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/35229#issuecomment-612579052) has completed - no new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/165fc96caa399b45b3a32be21bb40c2c86cff75a/pkgeval/by_hash/44fa722_vs_b8e9a9e/report.md). cc @maleadt

**Comment by maleadt** [comment]
The LLVM asserts build is broken on this branch (works with v1.4.0):

```
curl: (22) The requested URL returned error: 404 Not Found
make[1]: *** [/home/tim/Julia/julia/deps/llvm.mk:579: /home/tim/Julia/julia/deps/srccache/LLVM_full_assert.v8.0.1-3.x86_64-linux-gnu-cxx11.tar.gz] Error 22
```

https://github.com/JuliaBinaryWrappers/LLVM_full_assert_jll.jl/releases/download/LLVM_full_assert-v8.0.1+3/LLVM_full_assert.v8.0.1.x86_64-linux-gnu-cxx11.tar.gz doesn't exist whereas https://github.com/JuliaBinaryWrappers/LLVM_full_assert_jll.jl/releases/download/LLVM_full_assert-v8.0.1+0/LLVM_full_assert.v8.0.1.x86_64-linux-gnu-cxx11.tar.gz does. cc @vchuravy

**Pull request merged**